File: 2023/02/haskell/02.hs
{- stack script --resolver lts-14.20
                --package parsec -}
module Main where

import Text.Parsec hiding (parse)
import Data.Either

type Parser = Parsec String ()

parse :: Parsec String () a -> String -> Either ParseError a
parse p = runParser p () ""

number :: Parser Int
number = do
  int <- fmap read $ many1 digit
  return int

--------------------------------------------------------------------------------

data Game = Game Int [Draw] deriving Show
type Draw = (Color, Color, Color)
data Color = Red Int | Green Int | Blue Int deriving Show

colorToDraw :: Color -> Draw
colorToDraw (Red   r) = (Red r, Green 0, Blue 0)
colorToDraw (Green g) = (Red 0, Green g, Blue 0)
colorToDraw (Blue  b) = (Red 0, Green 0, Blue b)

emptyDraw :: Draw
emptyDraw = (Red 0, Green 0, Blue 0)

joinWith :: (Int -> Int -> Int ) -> Draw -> Draw -> Draw
joinWith f (Red r, Green g, Blue b) (Red r', Green g', Blue b') = (Red $ f r r', Green $ f g g', Blue $ f b b')

join :: Draw -> Draw -> Draw
join = joinWith (+)

--------------------------------------------------------------------------------

parseGame :: Parser Game
parseGame = do
  string "Game "
  id <- number
  string ": "
  draws <- sepBy parseDraw $ string "; "
  return $ Game id draws

parseColor :: Parser Color
parseColor = do
  cnt <- number
  many $ char ' '
  color <- string "red" <|> string "green" <|> string "blue"
  return $ case color of
    "red"   -> Red   cnt
    "green" -> Green cnt
    "blue"  -> Blue  cnt

parseDraw :: Parser Draw
parseDraw = do
  colors <- sepBy parseColor $ string ", "
  return $ foldl join emptyDraw $ map colorToDraw colors

--------------------------------------------------------------------------------

valid :: Draw -> Bool
valid (Red r, Green g, Blue b) = all id $ zipWith (<=) [r, g, b] [12, 13, 14]

draws :: Game -> [Draw]
draws (Game _ ds) = ds

gameId :: Game -> Int
gameId (Game id _) = id

getMax :: Game -> Game
getMax (Game id xs) = Game id $ (: []) $ foldl (joinWith max) emptyDraw xs

task1 :: [Game] -> Int
task1 gs = sum $ map gameId $ filter (valid . head . draws) $ map getMax gs

task2 :: [Game] -> Int
task2 gs = sum $ map (prod . head . draws . getMax) gs
  where
    prod :: Draw -> Int
    prod (Red r, Green g, Blue b) = r * g * b

main :: IO ()
main = do
  input <- fmap (rights . map (parse parseGame) . lines) getContents
  putStr "Task 1: "
  print $ task1 input
  putStr "Task 2: "
  print $ task2 input

File: 2023/02/rust/02.rs
use std::io;
use std::cmp::max;

#[derive(Clone)]
struct Game {
  id: i32,
  draws: Vec<[i32; 3]>,
}

fn parse_color(s: String) -> (usize, i32) {
  let mut spl = s.split(" ");
  let cnt = spl.next().unwrap()
               .parse()
               .ok().unwrap();
  let color = match spl.next().unwrap() {
    "red" => 0,
    "green" => 1,
    "blue" => 2,
    _ => 3,
  };
  (color, cnt)
}

fn parse_draw(s: String) -> [i32; 3] {
  let mut cs: [i32; 3] = [0, 0, 0];
  //let colors: Vec<(usize, i32)> = s.split(", ").collect::<Vec<&str>>().iter().map(|&s| parse_color(s.to_string())).collect();
  let v = s.split(", ").collect::<Vec<&str>>();
  let colors: Vec<(usize, i32)> = v.iter().map(|&s| parse_color(s.to_string())).collect();
  for (c, cnt) in colors {
    cs[c] += cnt;
  }
  cs
}

fn parse_game(s: String) -> Game {
  let mut game = s.split(": ");
  let id: i32 = game.next().unwrap().split(" ").nth(1).unwrap().parse().ok().unwrap();
  let draws: Vec<[i32; 3]> = game.next()
                                 .unwrap()
                                 .split("; ")
                                 .collect::<Vec<&str>>()
                                 .iter()
                                 .map(|s| parse_draw(s.to_string()))
                                 .collect();
  Game {
    id: id,
    draws: draws,
  }
}

fn task1(games: Vec<Game>) -> i32 {
  games.iter().map(|game| {
    let mut ms = [0, 0, 0];
    for draw in &game.draws {
      for i in 0..3 {
        ms[i] = max(ms[i], draw[i]);
      }
    }
    (game.id, ms)
  })
  .filter(|(_, ms)| ms.iter().zip([12, 13, 14].iter()).map(|(a, b)| a <= b).all(|x| x))
  .fold(0, |a, (id, _)| a + id)
}

fn task2(games: Vec<Game>) -> i32 {
  games.iter().map(|game| {
    let mut ms = [0, 0, 0];
    for draw in &game.draws {
      for i in 0..3 {
        ms[i] = max(ms[i], draw[i]);
      }
    }
    ms[0] * ms[1] * ms[2]
  }).fold(0, |a, b| a + b)
}

fn main() {
  let mut input: Vec<Game> = Vec::new();
  let mut s = String::new();
  while io::stdin().read_line(&mut s).expect("") > 0 {
    input.push(parse_game(s.trim().to_string()));
    s = String::new();
  }

  let t1 = task1(input.clone());
  println!("Task 1: {t1}");
  let t2 = task2(input.clone());
  println!("Task 2: {t2}");
}
